rfc/rfc-82/rfc-82.md

```diff
@@ -0,0 +1,254 @@
+| 2 | Not exists | S1 | S1 | No | S1 | Second commit, no schema evolution |
+| 3 | Not exists | S1 | S2 | Yes | N/A (throws exception) | No predefined schema, effectively concurrent schema definition. The design decision made here is to identify this as schema conflict for simplicity. It can also be broken down into cases where (1) S3 is evolvable from S2, (2) S3 is not evolvable from S2, for further optimization. |
+| 4 | S1 | S1 | S1 | No | S1 | No schema evolution |
+| 5 | S1 | S1 | S2 | No | S2 | Schema evolution in current transaction |
+| 6 | S1 | S2 | S1 | No | S2 | Backwards compatibility handles it |
+| 7 | S1 | S2 | S2 | No | S2 | Concurrent evolution to same schema |
+| 8 | S1 | S2 | S3 | Yes | N/A (throws exception) | Concurrent evolution to different schemas. Same as case 2, for future optimization we should consider evolvibility from S2 to S3 |
+
+For timeline graph of each case please refer appendix.
+
+Notes:
+- S1, S2, S3 represent different schemas. The proposed solution does not assume compability/evolvbility among those schemas.
+- 3 schemas to consider:
+  + The table schema from the last committed txn when the current txn starts.
+  + The table schema from the last committed txn when the current txn validates.
+  + The table schema that the current txn uses.
+- "Not exists" means the table is empty and table schema is not set.
+
+## Implementation
+
+The proposed implementation involves the following key changes:
+
+1. Enhance the `TransactionUtils` class to include schema conflict detection:
+   - Add a new method `abortTxnOnConcurrentSchemaEvolution` to check for schema conflicts.
+
+2. Schema conflict detection logic:
+   - Follow the graph as explained above.
+
+## Rollout/Adoption Plan
+
+1. Impact on existing users:
+   - Users leveraging multi-writer capabilities with schema evolution will benefit from improved consistency.
+   - No breaking changes for users not using concurrent schema evolution.
+
+2. Phasing out older behavior:
+   - The code is not protected by a feature flag, once it is merged the behavior is adopted in the next release.
+
+3. Migration:
+   - No special migration tools are required.
+   - Users should update to the latest Hudi version to benefit from this enhancement.
+
+
+## Test Plan
+
+The implementation includes comprehensive test cases in the `TestHoodieClientMultiWriter`:
+
+1. Test scenarios:
+   - Covered by the table in the previous section.
+
+2. Test variations:
+   - Both COPY_ON_WRITE and MERGE_ON_READ table types
+
+# Appendix
+
+## Timeline of concurrency schema evolution case
+
+**Scenario 1**
+
+
+Should Conflict?: No
+```
+Time ---------------------------------------------------------------------------------->
+
+Txn1:    [ Start Txn1 ]------------------------------[ Validate ]--------------------[ Commit Txn1 ]
+            Uses Schema: S1                                                     Writes Schema: S1
```

Comment:
For clarity and consistency of writing, use the same defined terms throughout (same for other scenarios)
```suggestion
            Uses writer schema: S1                                          Writes table schema in commit metadata: S1
```

Reply:
done